cluster-autoscaler/proposals/provisioning-request.md

```diff
@@ -0,0 +1,232 @@
+
+### Provisioning Request CRD
+
+The following code snippets assume [kubebuilder](https://book.kubebuilder.io/)
+is ued to generate the CRD:
+
+```go
+// ProvisioningRequest is a way to express additional capacity
+// that we would like to provision in the cluster. Cluster Autoscaler
+// can use this information in its calculations and signal if the capacity
+// is available in the cluster or actively add capacity if needed.
+type ProvisioningRequest struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	//
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Spec contains specification of the ProvisioningRequest object.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status.
+	//
+	// +kubebuilder:validation:Required
+	Spec ProvisioningRequestSpec `json:"spec" protobuf:"bytes,2,name=spec"`
+	// Status of the ProvisioningRequest. CA constantly reconciles this field.
+	//
+	// +optional
+	Status ProvisioningRequestStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// ProvisioningRequestList is a object for list of ProvisioningRequest.
+type ProvisioningRequestList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	//
+	// +optional
+	metav1.ListMeta `json:"metadata" protobuf:"bytes,1,opt,name=metadata"`
+	// Items, list of ProvisioningRequest returned from API.
+	//
+	// +optional
+	Items []ProvisioningRequest `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// ProvisioningRequestSpec is a specification of additional pods for which we
+// would like to provision additional resources in the cluster.
+type ProvisioningRequestSpec struct {
+	// PodSets lists groups of pods for which we would like to provision
+	// resources.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
+	PodSets []PodSet `json:"podSets" protobuf:"bytes,1,rep,name=podSets"`
+
+	// OperationMode describes the different modes of provisioning the resources.
+	// Supported values:
+	// * GENERIC_CHECK_CAPACITY - check if current cluster state can fullfil this request
+	// * GENERIC_ATOMIC_SCALE_UP - provision the resources in an atomic manner
+    // * ... - potential other modes that are specific to the cloud providers
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
+	OperationMode string `json:"operationMode" protobuf:"bytes,2,name=operationMode"`
+
+	// AdditionalParameters contains all other parameters custom modes may require.
+	//
+	// +optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
+	AdditionalParameters map[string]string `json:"additionalParameters" protobuf:"bytes,3,name=additionalParameters"`
```

Comment:
If we switched to ProvisioningClass, this would go to the ProvisioningClass object and would be set by the system administrator, not the caller, who may have less knowledge of what can possibly be set. On the other hand, some workload specific parameters may not fit well into ProvisioningClass (like expected workload duration, for instance). WDYT?

Reply:
I think addition of the ProvisioningClass CRD would result in some parameters being passed in ProvisioningRequest and others in ProvisioningClass. In my opinion this obscures the parameters users may be passing to CA.
I would restrain from adding yet another CRD as for now we do not have a clear use-case for parameters admins may want to pass through it.

Reply:
+1 to @kisieland , additional CRD seems premature before we have some actual admin-level parameters. As for listing, we will only have a few classes (maybe 2-4 depending how many cloud providers add), so it's easy to list in the documentation.

Reply:
We have 30+ cloud providers in cluster autoscaler. Each of them may introduce a vendor-specific modes/classes for provisioning the resources. In the worst/best case scenario (depending how you look at it), if each of them introduces 4 we will have 120 possible values, that I would rather don't want to list in the documentation/check in validating code. 

We don't need to have ProvisioningClass with all bells and whistles from day 1, but this direction should probably be mentioned in the doc, along with the intermediate steps (like sole string with no validation), and how we get from the intermediate solution to the final one.

Reply:
Added a section with a possible future expansion and an example of how the ProvisioningClass CRD would look like.

Reply:
We can make it inmutable, while allowing the annotations to change.
Also this will make it clear when a user is providing any additional data.
But if you prefer to pipe it through annotations we should be able to d o it as well.